Metaswitch/clearwater-cassandra
# Cass 2.1.15

I think the most important thing in this pull request, is to check that any of the options that been removed/changed are not things that we think are important. 
## Main config changes

Biggest apparent changes are in the env.sh and yaml files.
For these, i took the files from cassandra 2.1.15, and applied the changes that we had previously applied to the 2.0.14 files. On top of those, i then made a few further changes.

The specific changes i have made, as opposed to diffs between cassandra versions, are:
In env.sh:
- Addded `JVM_OPTS="$JVM_OPTS -Dcassandra.ignore_rack=true -Dcassandra.ignore_dc=true"`, as the new version is stricter about overwriting these values, which we do on startup.

In cassandra.yaml:
- Uncommented lines setting directory paths, as casssandra has changed the default location. This included:
  - data_file_directories
  - commitlog_directory
  - saved_caches_directory
- Set the template `broadcast_rpc_address` to 127.0.0.1 so it doesn't break on startup, and overwrite this in the plugin to be the node IP/same as `listen_address`
## Other changes:
- Moved the `touch force_cassandra_yaml_refresh` to a preinst file, rather than the postinst, as this allows us to run it only on upgrade. We don't want to run it on fresh installs, as there are two side-effects:
  - If adding a new node to a cluster, Cassandra 2.1.15 doesn't like the multiple restarts and rejoining clusters, and complains that it is already in the cluster it is attempting to join.
  - More importantly, the previous behaviour was that a new node would join the cluster, have time to get data, and then be destructively restarted (removing data) before rejoining again. If you added <replication-factor> number of nodes or more in one go, there is the possibility that they would all join, get given data, and be the sole possessors of that data, and then hit destructive restart, losing those keys completely.

Cassandra have removed the following options from the .yaml file, which i assume is fine, but may be worth checking. We hadn't ever changed them, so unless something stands out, should be fine as it is:
- memtable_flush_queue_size: 4
- in_memory_compaction_limit_in_mb: 64
- multithreaded_compaction: false
- compaction_preheat_key_cache: true
- preheat_kernel_page_cache: false
## Testing:

In kicking the tires, since working through all the install/setup issues, i have run through the following scenarios, and seen no problems:
- Set up a 2.0.14 cluster, provision data into cassandra, upgrade nodes to 2.1.15, check data still present.
- Set up a 2.1.15 cluster, run all methods of bulk provisioning. Use cqlsh to check data in place correctly.

The correct upgrade process has not yet been figured out, so we can't just merge away straight off.
